[*] Reduce contract size

TuDo1403 opened the pull request:
### Description
This PR reduces overall contract sizes using following techniques:
- Convert require string to revert custom error
- Uncheck for loop/while loop iteration
- Setting bytecodehash to 'none'
- Uncheck incrementation
- Has ... Contract merging
- Assemblify hash struct operation to avoid memory expansion
- Move reused modifier logic to internal function
### Contract changes

The table below shows the following info:
- **Logic**: the logic is changed.
- **ABI**: the ABI is changed.
- **Init data**: new storage field is declared and needs initializing.
- **Dependent**: needs to be changed due to changes in other contracts.

| **Contract name** | **Logic** | **ABI** | **Init data** | **Dependent** |
|-------------------|:---------:|:-------:|:-------------:|:-------------:|
| BridgeTracking    |      x     |    x     |      x         |               |
| Maintenance       |     x      |      x   |         x      |               |
| SlashIndicator    |     x      |      x   |       x        |               |
| Staking           |      x     |   x     |          x     |               |
| StakingVesting    |     x      |    x     |      x         |               |
| ValidatorSet      |     x      |    x     |       x        |               |
| MainchainGovernanceAdmin   |           |    x     |               |               |
| MainchainGateway   |           |    x     |               |               |
| MainchainPauseEnforcer   |           |    x     |               |               |
| RoninGovernanceAdmin   |           |    x     |               |               |
| RoninGateway   |           |    x     |               |               |
| RoninPauseEnforcer   |           |    x     |               |               |

### Checklist
- [x] I have clearly commented on all the main functions following the [NatSpec Format](https://docs.soliditylang.org/en/v0.8.0/natspec-format.html)
- [x] The box that allows repo maintainers to update this PR is checked
- [x] I tested locally to make sure this feature/fix works

Commit:
[*] Reduce code size: uncheck for loop iterations (#229)

* feat: reduce overall code size by unchecking iteration

* fix: skip unchecking for nondeterministic for loop

* Update hardhat.config.ts

Co-authored-by: Bao <quocbao300@gmail.com>

* chore: update yarn.lock

* chore: move log -> logs

* feat: optimizer run 10 for RoninValidatorSet

---------

Co-authored-by: Bao <quocbao300@gmail.com>

Commit:
[*]: Reduce code size: Revert custom error (#232)

* feat: reduce overall code size by unchecking iteration

* fix: skip unchecking for nondeterministic for loop

* Update hardhat.config.ts

Co-authored-by: Bao <quocbao300@gmail.com>

* chore: update yarn.lock

* chore: move log -> logs

* feat: optimizer run 10 for RoninValidatorSet

* feat: convert all require strings to revert custom errors

* fix: fix branching error

* fix: remove redundant file

* format: rename ErEmptyArrayLength error

* format: rename ErrEmptyArrayLength error

* format: add comment for ErrorHandler lib

* feat: provide more info on revert ErrUnauthorized

* format: comment for error code

* fix: convert hardcoded sighash to function for better readability

* feat: inline revert

* feat: add more informative error

* feat: update custom errors

* fix: rename roles

* fix: rename roles and comment index

---------

Co-authored-by: Bao <quocbao300@gmail.com>

Commit:
[*] Reduce codesize: Modifier code size reduce (#233)

* feat: reduce overall code size by unchecking iteration

* fix: skip unchecking for nondeterministic for loop

* Update hardhat.config.ts

Co-authored-by: Bao <quocbao300@gmail.com>

* chore: update yarn.lock

* chore: move log -> logs

* feat: optimizer run 10 for RoninValidatorSet

* feat: convert all require strings to revert custom errors

* feat: move reused modifier body to internal function

* fix: fix branching error

* Update contracts/libraries/Proposal.sol

Co-authored-by: Bao <quocbao300@gmail.com>

* fix: remove redundant file

* format: rename ErEmptyArrayLength error

* format: rename ErrEmptyArrayLength error

* format: add comment for ErrorHandler lib

* feat: provide more info on revert ErrUnauthorized

* format: comment for error code

* fix: convert hardcoded sighash to function for better readability

* feat: inline revert

* feat: add more informative error

* feat: update custom errors

* feat: move reused modifier body to internal function

* Update contracts/libraries/Proposal.sol

Co-authored-by: Bao <quocbao300@gmail.com>

* tmp: tmp commit

* format: update format

* format: fix function name

* Delete .yarnrc.yml

* Delete .gitattributes

* Delete .editorconfig

* Delete yarn-1.22.19.cjs

* format: group import

---------

Co-authored-by: Bao <quocbao300@gmail.com>

Commit:
chore: gitignore for yarn berry

Commit:
feat: bump hardhat-contract-sizer@2.8.0

ducthotran2010 commented on contracts/extensions/GovernanceAdmin.sol:
```diff
@@ -5,16 +5,19 @@ import "../extensions/sequential-governance/CoreGovernance.sol";
 import "../interfaces/IRoninTrustedOrganization.sol";
+import "../libraries/ErrorHandler.sol";
+
+import "../libraries/ErrorHandler.sol";
```
```suggestion
```

TuDo1403 replied:
removed.

ducthotran2010 commented on contracts/extensions/GovernanceAdmin.sol:
```diff
@@ -76,7 +79,8 @@ abstract contract GovernanceAdmin is CoreGovernance, HasRoninTrustedOrganization
     (bool _success, bytes memory _returndata) = _proxy.staticcall(abi.encodeWithSelector(_selector));
-    if (!_success) revert ErrProxyCallFailed(_selector);
+    _success.handleRevert(_selector, _returndata);
+
```
```suggestion
```

ducthotran2010 commented on contracts/extensions/GovernanceAdmin.sol:
```diff
@@ -5,16 +5,19 @@ import "../extensions/sequential-governance/CoreGovernance.sol";
-
   modifier onlySelfCall() {
-    require(msg.sender == address(this), "GovernanceAdmin: only allowed self-call");
+    _onlySelfCall();
```
```suggestion
    _requireSelfCall();
```

TuDo1403 replied:
fixed.

ducthotran2010 commented on contracts/extensions/GovernanceAdmin.sol:
```diff
@@ -145,7 +149,11 @@ abstract contract GovernanceAdmin is CoreGovernance, HasRoninTrustedOrganization
     return abi.decode(_returndata, (uint256));
   }
+
+  function _onlySelfCall() internal view virtual {
```
```suggestion
  function _requireSelfCall() internal view virtual {
```

TuDo1403 replied:
resolved.

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BOsGovernanceProposal.sol:
```diff
@@ -7,6 +7,11 @@ import "../../interfaces/IRoninGovernanceAdmin.sol";
+   * @dev Error indicating that the order of signers is invalid.
+   */
+  error ErrInvalidSignerOrder();
+
```
Place error defines after `using` keywords

TuDo1403 replied:
fixed.

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BOsGovernanceRelay.sol:
```diff
@@ -58,20 +64,21 @@ abstract contract BOsGovernanceRelay is SignatureConsumer, VoteStatusConsumer {
       _signers[_i] = ECDSA.recover(_digest, _sig.v, _sig.r, _sig.s);
-      require(_lastSigner < _signers[_i], "BOsGovernanceRelay: invalid order");
+      if (_lastSigner >= _signers[_i]) revert ErrInvalidOrder(msg.sig);
+
```
```suggestion
```

ducthotran2010 commented on contracts/extensions/collections/HasProxyAdmin.sol:
```diff
@@ -3,12 +3,14 @@ pragma solidity ^0.8.0;
 
   modifier onlyAdmin() {
-    require(msg.sender == _getAdmin(), "HasProxyAdmin: unauthorized sender");
+    _onlyAdmin();
```
```suggestion
    _requireAdmin();
```

ducthotran2010 commented on contracts/extensions/collections/HasProxyAdmin.sol:
```diff
@@ -18,4 +20,8 @@ abstract contract HasProxyAdmin {
     return StorageSlot.getAddressSlot(_ADMIN_SLOT).value;
   }
+
+  function _onlyAdmin() internal view {
```
```suggestion
  function _requireAdmin() internal view {
```

ducthotran2010 commented on contracts/extensions/collections/HasProxyAdmin.sol:
```diff
@@ -3,12 +3,14 @@ pragma solidity ^0.8.0;
 
 import "@openzeppelin/contracts/utils/StorageSlot.sol";
 
+import "../../libraries/Errors.sol";
```
Weird grouping

TuDo1403 replied:
resolved.

ducthotran2010 commented on contracts/extensions/forwarder/Forwarder.sol:
```diff
@@ -4,7 +4,16 @@ pragma solidity ^0.8.0;
 
 import "@openzeppelin/contracts/access/AccessControlEnumerable.sol";
 
+import { ErrorHandler } from "../../libraries/ErrorHandler.sol";
```
Weird import grouping

TuDo1403 replied:
resolved.

ducthotran2010 commented on contracts/extensions/forwarder/Forwarder.sol:
```diff
@@ -4,7 +4,16 @@ pragma solidity ^0.8.0;
+   */
+  error ErrInvalidForwardValue();
+
+  using ErrorHandler for bool;
```
```suggestion
  using ErrorHandler for bool;

  /**
   * @dev Error thrown when an invalid forward value is provided.
   */
  error ErrInvalidForwardValue();
```

ducthotran2010 commented on contracts/extensions/forwarder/Forwarder.sol:
```diff
@@ -22,8 +31,12 @@ contract Forwarder is AccessControlEnumerable {
+    for (uint _i = 0; _i < _targets.length; ) {
       _setupRole(TARGET_ROLE, _targets[_i]);
+
+      unchecked {
+        ++_i;
+      }
```
Should we make it as an inline uncheck

TuDo1403 replied:
unchecked block cannot be inlined.

ducthotran2010 commented on contracts/libraries/Errors.sol:
CommonErrors.sol

TuDo1403 replied:
updated.

ducthotran2010 commented on contracts/libraries/GlobalProposal.sol:
```diff
@@ -77,13 +86,18 @@ library GlobalProposal {
+      }
+
+      unchecked {
+        ++_i;
```
Nani?
```suggestion
      unchecked {
        ++_i;
```

TuDo1403 replied:
resolved.

ducthotran2010 commented on contracts/libraries/AddressArrayUtils.sol:
```diff
@@ -2,6 +2,12 @@
 
 pragma solidity ^0.8.0;
 
+/**
+ * @dev Error thrown when a duplicated element is detected in an array.
+ * @param msgSig The function signature that invoke the error.
+ */
+error ErrDuplicated(bytes4 msgSig);
```
Should we move this error into the lib?

TuDo1403 replied:
only RoninTrustedOrganization.sol and CandidateStaking.sol which import AddressArrayUtils lib used this error.

ducthotran2010 commented on contracts/libraries/BridgeOperatorsBallot.sol:
```diff
@@ -3,7 +3,14 @@ pragma solidity ^0.8.0;
 
 import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
 
+import "../libraries/Errors.sol";
```
Weird grouping

TuDo1403 replied:
resolved.

ducthotran2010 commented on contracts/libraries/Proposal.sol:
```diff
@@ -1,7 +1,19 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.0;
 
+import "../libraries/Errors.sol";
```
Could it be:
```solidity
import { UsedErrors,... } from "../libraries/Errors.sol";
```

TuDo1403 replied:
updated.

ducthotran2010 commented on contracts/libraries/Roles.sol:
```diff
@@ -0,0 +1,23 @@
+pragma solidity ^0.8.0;
+
+enum Roles {
+  ADMIN, // 0
```
1. Reserve a slot for unknown role
2. Only use prefix `s` for arrays
```suggestion
enum Role {
  UNKNOWN,
  ADMIN,
```

TuDo1403 replied:
renamed.

ducthotran2010 commented on contracts/mainchain/MainchainGovernanceAdmin.sol:
```diff
@@ -8,7 +8,11 @@ import "../extensions/TransparentUpgradeableProxyV2.sol";
 import "../extensions/GovernanceAdmin.sol";
 import "../interfaces/IBridge.sol";
 
+import "../libraries/ErrorHandler.sol";
```
Weird grouping

TuDo1403 replied:
fixed.

ducthotran2010 commented on contracts/ronin/gateway/BridgeTracking.sol:
```diff
@@ -44,10 +44,14 @@ contract BridgeTracking is HasBridgeContract, HasValidatorContract, Initializabl
       return;
     }
-    _;
   }
```
Does it return for the whole trxn?

TuDo1403 replied:
fixed.

ducthotran2010 commented on contracts/ronin/staking/BaseStaking.sol:
```diff
@@ -33,25 +33,41 @@ abstract contract BaseStaking is
     _;
   }
 
+  function _onlyEmptyValue() private view {
```
```suggestion
  function _requireValue() private view {
```

TuDo1403 replied:
resolved.

nxqbao committed:
fix rebase issue

ducthotran2010 commented on contracts/ronin/validator/CoinbaseExecution.sol:
```diff
@@ -30,10 +30,14 @@ abstract contract CoinbaseExecution is
     _;
   }
 
+  function _onlyCoinbase() private view {
```
```suggestion
  function _requireCoinbase() private view {
```

TuDo1403 replied:
resolved.

nxqbao committed:
Fix package issue. Upgrade node version for CI.

nxqbao committed:
add contract size log

ducthotran2010 commented on logs/contract-code-sizes-1685028054516.log:
Do we need this file?

TuDo1403 replied:
deleted.

ducthotran2010 reviewed (changes requested):
Most of the issues come from:
- Confused `import` grouping
- Funciton naming
- Invalid trailing line
- Invalid keyword orders (`error` -> `using`)

Some suggestions that should be applied:
- Inline uncheck for-loop

Commit:
format: rename internal fn

Commit:
format: rename internal fn

Commit:
format: remove duplicated import

Commit:
format: reorder import

Commit:
chore: reduce unused log file

Commit:
format: rename internal fn

Commit:
Update contracts/extensions/GovernanceAdmin.sol

Co-authored-by: Duc Tho Tran <ducthotran2010@gmail.com>

Commit:
format: fix weird grouping

Commit:
fix: fix bug not return whole tx if using internal fn

Commit:
format: rename internal fn

Commit:
Update contracts/extensions/bridge-operator-governance/BOsGovernanceRelay.sol

Co-authored-by: Duc Tho Tran <ducthotran2010@gmail.com>

Commit:
format: rename internal fn

Commit:
format: fix weird grouping

Commit:
format: fix weird grouping

Commit:
Update contracts/extensions/forwarder/Forwarder.sol

Co-authored-by: Duc Tho Tran <ducthotran2010@gmail.com>

Commit:
feat: rename Errors.sol -> CommonErrors.sol

Commit:
format: import only used error from lib

Commit:
format: rename enum Roles -> Role

Commit:
fix: fix error rename

Commit:
format: fix weird import grouping

Commit:
fix: remove duplicated for loop increment

Commit:
feat: uncheck while loop iteration

Commit:
feat: update contract size log

Commit:
[*] Reduce code size: remove bytecodehash (#245)

* feat: set bytecodehash to none to avoid nondeterministic bytecode and reduce overall code size by 0.04kb

* Update hardhat.config.ts

Co-authored-by: Duc Tho Tran <ducthotran2010@gmail.com>

---------

Co-authored-by: Duc Tho Tran <ducthotran2010@gmail.com>

Commit:
format: format code

ducthotran2010 commented:
This PR has some issues related to prettier

Commit:
[*] Reduce Contract Size: uncheck all incrementation (#246)

feat: uncheck increment

Commit:
[*] Reduce code size: efficient hashing to avoid memory expansion (#241)

* [*]: Reduce code size: Revert custom error (#232)

* feat: reduce overall code size by unchecking iteration

* fix: skip unchecking for nondeterministic for loop

* Update hardhat.config.ts

Co-authored-by: Bao <quocbao300@gmail.com>

* chore: update yarn.lock

* chore: move log -> logs

* feat: optimizer run 10 for RoninValidatorSet

* feat: convert all require strings to revert custom errors

* fix: fix branching error

* fix: remove redundant file

* format: rename ErEmptyArrayLength error

* format: rename ErrEmptyArrayLength error

* format: add comment for ErrorHandler lib

* feat: provide more info on revert ErrUnauthorized

* format: comment for error code

* fix: convert hardcoded sighash to function for better readability

* feat: inline revert

* feat: add more informative error

* feat: update custom errors

* fix: rename roles

* fix: rename roles and comment index

---------

Co-authored-by: Bao <quocbao300@gmail.com>

* [*] Reduce codesize: Modifier code size reduce (#233)

* feat: reduce overall code size by unchecking iteration

* fix: skip unchecking for nondeterministic for loop

* Update hardhat.config.ts

Co-authored-by: Bao <quocbao300@gmail.com>

* chore: update yarn.lock

* chore: move log -> logs

* feat: optimizer run 10 for RoninValidatorSet

* feat: convert all require strings to revert custom errors

* feat: move reused modifier body to internal function

* fix: fix branching error

* Update contracts/libraries/Proposal.sol

Co-authored-by: Bao <quocbao300@gmail.com>

* fix: remove redundant file

* format: rename ErEmptyArrayLength error

* format: rename ErrEmptyArrayLength error

* format: add comment for ErrorHandler lib

* feat: provide more info on revert ErrUnauthorized

* format: comment for error code

* fix: convert hardcoded sighash to function for better readability

* feat: inline revert

* feat: add more informative error

* feat: update custom errors

* feat: move reused modifier body to internal function

* Update contracts/libraries/Proposal.sol

Co-authored-by: Bao <quocbao300@gmail.com>

* tmp: tmp commit

* format: update format

* format: fix function name

* Delete .yarnrc.yml

* Delete .gitattributes

* Delete .editorconfig

* Delete yarn-1.22.19.cjs

* format: group import

---------

Co-authored-by: Bao <quocbao300@gmail.com>

* feat: efficient hashing

* [CI] Test for `reduce-contract-size` branch (#244)

add ci for reduce-contract-size

* feat: add contract size log

* feat: update contract size log

* format: rename parameter for new code convention

* format: format code

* fix: remove storage layout log

* fix: delete storage layout

* fix: delete yarn lock

* feat: remove changes in yarn.lock

* revert change in package.json & yarn.lock

---------

Co-authored-by: Bao <quocbao300@gmail.com>
Co-authored-by: Duc Tho Tran <ducthotran2010@gmail.com>

Commit:
Merge branch 'dev' into reduce-contract-size

Commit:
[*] Reduce code size: Remove duplicated authorized contract getter/settter logics utilizing only 1 abstract class (#242)

* [*]: Reduce code size: Revert custom error (#232)

* feat: reduce overall code size by unchecking iteration

* fix: skip unchecking for nondeterministic for loop

* Update hardhat.config.ts

Co-authored-by: Bao <quocbao300@gmail.com>

* chore: update yarn.lock

* chore: move log -> logs

* feat: optimizer run 10 for RoninValidatorSet

* feat: convert all require strings to revert custom errors

* fix: fix branching error

* fix: remove redundant file

* format: rename ErEmptyArrayLength error

* format: rename ErrEmptyArrayLength error

* format: add comment for ErrorHandler lib

* feat: provide more info on revert ErrUnauthorized

* format: comment for error code

* fix: convert hardcoded sighash to function for better readability

* feat: inline revert

* feat: add more informative error

* feat: update custom errors

* fix: rename roles

* fix: rename roles and comment index

---------

Co-authored-by: Bao <quocbao300@gmail.com>

* [*] Reduce codesize: Modifier code size reduce (#233)

* feat: reduce overall code size by unchecking iteration

* fix: skip unchecking for nondeterministic for loop

* Update hardhat.config.ts

Co-authored-by: Bao <quocbao300@gmail.com>

* chore: update yarn.lock

* chore: move log -> logs

* feat: optimizer run 10 for RoninValidatorSet

* feat: convert all require strings to revert custom errors

* feat: move reused modifier body to internal function

* fix: fix branching error

* Update contracts/libraries/Proposal.sol

Co-authored-by: Bao <quocbao300@gmail.com>

* fix: remove redundant file

* format: rename ErEmptyArrayLength error

* format: rename ErrEmptyArrayLength error

* format: add comment for ErrorHandler lib

* feat: provide more info on revert ErrUnauthorized

* format: comment for error code

* fix: convert hardcoded sighash to function for better readability

* feat: inline revert

* feat: add more informative error

* feat: update custom errors

* feat: move reused modifier body to internal function

* Update contracts/libraries/Proposal.sol

Co-authored-by: Bao <quocbao300@gmail.com>

* tmp: tmp commit

* format: update format

* format: fix function name

* Delete .yarnrc.yml

* Delete .gitattributes

* Delete .editorconfig

* Delete yarn-1.22.19.cjs

* format: group import

---------

Co-authored-by: Bao <quocbao300@gmail.com>

* feat: merge collections contract to one contract

* feat: apply HasContract to other contracts

* chore: remove unused contracts

* [CI] Test for `reduce-contract-size` branch (#244)

add ci for reduce-contract-size

* feat: add contract size log

* chore: rename deprecated values

* chore: rename custom slot

* feat: change ErrZeroCodeContract parameter

* feat: update contract size log

* Update contracts/ronin/validator/CoinbaseExecution.sol

Co-authored-by: Bao <quocbao300@gmail.com>

* Update contracts/interfaces/collections/IHasContract.sol

Co-authored-by: Bao <quocbao300@gmail.com>

* feat: reserve 0 index for enum Roles

* chore: rename EMERGENCY_PAUSER -> PAUSE_ENFORCER_CONTRACT

* feat: arrange role contract or eoa by parity order

* [*]: Reduce code size: Revert custom error (#232)

* feat: reduce overall code size by unchecking iteration

* fix: skip unchecking for nondeterministic for loop

* Update hardhat.config.ts

Co-authored-by: Bao <quocbao300@gmail.com>

* chore: update yarn.lock

* chore: move log -> logs

* feat: optimizer run 10 for RoninValidatorSet

* feat: convert all require strings to revert custom errors

* fix: fix branching error

* fix: remove redundant file

* format: rename ErEmptyArrayLength error

* format: rename ErrEmptyArrayLength error

* format: add comment for ErrorHandler lib

* feat: provide more info on revert ErrUnauthorized

* format: comment for error code

* fix: convert hardcoded sighash to function for better readability

* feat: inline revert

* feat: add more informative error

* feat: update custom errors

* fix: rename roles

* fix: rename roles and comment index

---------

Co-authored-by: Bao <quocbao300@gmail.com>

* [*] Reduce codesize: Modifier code size reduce (#233)

* feat: reduce overall code size by unchecking iteration

* fix: skip unchecking for nondeterministic for loop

* Update hardhat.config.ts

Co-authored-by: Bao <quocbao300@gmail.com>

* chore: update yarn.lock

* chore: move log -> logs

* feat: optimizer run 10 for RoninValidatorSet

* feat: convert all require strings to revert custom errors

* feat: move reused modifier body to internal function

* fix: fix branching error

* Update contracts/libraries/Proposal.sol

Co-authored-by: Bao <quocbao300@gmail.com>

* fix: remove redundant file

* format: rename ErEmptyArrayLength error

* format: rename ErrEmptyArrayLength error

* format: add comment for ErrorHandler lib

* feat: provide more info on revert ErrUnauthorized

* format: comment for error code

* fix: convert hardcoded sighash to function for better readability

* feat: inline revert

* feat: add more informative error

* feat: update custom errors

* feat: move reused modifier body to internal function

* Update contracts/libraries/Proposal.sol

Co-authored-by: Bao <quocbao300@gmail.com>

* tmp: tmp commit

* format: update format

* format: fix function name

* Delete .yarnrc.yml

* Delete .gitattributes

* Delete .editorconfig

* Delete yarn-1.22.19.cjs

* format: group import

---------

Co-authored-by: Bao <quocbao300@gmail.com>

* format: rename internal fn

* feat: update contract code sizes log

* feat: update contract size log

* format: format code

* feat: rename HasContract -> HasContracts

---------

Co-authored-by: Bao <quocbao300@gmail.com>

Commit:
feat: preserve storage slot layouts

Commit:
feat: update contract size log

nxqbao commented on contracts/libraries/DeprecatedSlots.sol:
```diff
@@ -0,0 +1,45 @@
+ * @dev These abstract contracts are deprecated and should not be used in new implementations.
+ * They provide functionality related to various aspects of a smart contract but have been marked
+ * as deprecated to indicate that they are no longer actively maintained or recommended for use.
+ * @notice The reason for these contracts is for preserving the slots for already deployed contracts.
```
```suggestion
 * The purpose of these contracts is to preserve the slots for already deployed contracts.
```

nxqbao commented on contracts/libraries/DeprecatedSlots.sol:
```diff
@@ -0,0 +1,45 @@
+pragma solidity ^0.8.0;
+
+/**
+ * @title Deprecated Contracts
```
Should move this file to `contracts/utils/DeprecatedSlots.sol`, since these contracts are not library.

nxqbao commented on contracts/libraries/DeprecatedSlots.sol:
```diff
@@ -0,0 +1,45 @@
+ * @notice The reason for these contracts is for preserving the slots for already deployed contracts.
+ */
+contract HasSlashIndicatorDeprecated {
+  address private ____deprecatedSI;
```
Please add `@custom:deprecated` NatSpecs comment for this and other relevant variables.

Commit:
format: update comment

nxqbao commented on contracts/libraries/GlobalProposal.sol:
```diff
@@ -77,13 +81,15 @@ library GlobalProposal {
         _detail.targets[_i] = _roninTrustedOrganizationContract;
-      } else {
-        revert("GlobalProposal: unsupported target");
+      } else revert ErrUnsupportedTarget();
```
Should add proposal hash and target number for ease when debugging, since this error is in a loop.

nxqbao commented on contracts/libraries/Proposal.sol:
```diff
@@ -85,20 +100,24 @@ library Proposal {
-    for (uint256 _i = 0; _i < _proposal.targets.length; ++_i) {
-      require(gasleft() > _proposal.gasAmounts[_i], "Proposal: insufficient gas");
+    for (uint256 _i = 0; _i < _proposal.targets.length; ) {
+      if (gasleft() <= _proposal.gasAmounts[_i]) revert ErrInsufficientGas();
```
This error is in a loop, please add proposal hash into the error.

nxqbao commented on contracts/libraries/Role.sol:
```diff
@@ -0,0 +1,32 @@
+ * @dev Sort role by contract or eoa in parity order.
+ * @notice 1st bit is reserved
+ * @notice EOA is even (*0), contract is odd (*1)
+ */
```
```suggestion
/**
 * @notice Sort role by contract or EOA in parity order.
 * @dev The 1st (the right bit) most is reserved. EOA addresses are even (`*0`), contract addresses are odd (`*1`)
 */
```

nxqbao commented on contracts/libraries/Role.sol:
```diff
@@ -0,0 +1,32 @@
+  RESERVE_0, // 18
+  STAKING_CONTRACT, // 19
+  RESERVE_1, // 20
+  RONIN_TRUSTED_ORGANIZATION_CONTRACT // 21
```
```suggestion
   /*  0 */ UNKNOWN_0,
   /*  1 */ UNKNOWN_1,
   /*  2 */ ADMIN,
   /*  3 */ PAUSE_ENFORCER_CONTRACT,
   /*  4 */ COINBASE,
   /*  5 */ BRIDGE_CONTRACT,
   /*  6 */ GOVERNOR,
   /*  7 */ BRIDGE_TRACKING_CONTRACT,
   /*  8 */ CANDIDATE_ADMIN,
   /*  9 */ GOVERNANCE_ADMIN_CONTRACT,
   /* 10 */ WITHDRAWAL_MIGRATOR,
   /* 11 */ MAINTENANCE_CONTRACT,
   /* 12 */ BRIDGE_OPERATOR,
   /* 13 */ SLASH_INDICATOR_CONTRACT,
   /* 14 */ BLOCK_PRODUCER,
   /* 15 */ STAKING_VESTING_CONTRACT,
   /* 16 */ VALIDATOR_CANDIDATE,
   /* 17 */ VALIDATOR_CONTRACT,
   /* 18 */ RESERVED_0,
   /* 19 */ STAKING_CONTRACT,
   /* 20 */ RESERVED_1,
   /* 21 */ RONIN_TRUSTED_ORGANIZATION_CONTRACT
```

nxqbao commented on contracts/libraries/Token.sol:
```diff
@@ -7,10 +7,31 @@ import "@openzeppelin/contracts/utils/Strings.sol";
+  /**
+   * @dev Error indicating that an unsupported standard is encountered.
+   */
+  error ErrUnsupportedStandard();
```
```suggestion
  /// @dev Error indicating that the provided information is invalid.
  error ErrInvalidInfo();

  /// @dev Error indicating that the minting of ERC20 tokens has failed.
  error ErrERC20MintingFailed();

  /// @dev Error indicating that the minting of ERC721 tokens has failed.
  error ErrERC721MintingFailed();

  /// @dev Error indicating that an unsupported standard is encountered.
  error ErrUnsupportedStandard();
```

nxqbao commented on contracts/ronin/gateway/RoninGatewayV2.sol:
```diff
@@ -52,16 +48,16 @@ contract RoninGatewayV2 is
+  /// @custom:deprecated previously `_validatorContract` (non-zero value)
+  address private ____deprecated0;
+  /// @custom:deprecated previously `_bridgeTrackingContract` (non-zero value)
+  address private ____deprecated1;
```
```suggestion
  /// @custom:deprecated Previously `_validatorContract` (non-zero value)
  address private ____deprecated0;
  /// @custom:deprecated Previously `_bridgeTrackingContract` (non-zero value)
  address private ____deprecated1;
```

nxqbao commented on contracts/ronin/gateway/RoninGatewayV2.sol:
```diff
@@ -52,16 +48,16 @@ contract RoninGatewayV2 is
 
-  /// @dev The trusted organization contract
-  IRoninTrustedOrganization internal _trustedOrgContract;
+  /// @custom:deprecated previously `_trustedOrgContract` (non-zero value)
```
```suggestion
  /// @custom:deprecated Previously `_trustedOrgContract` (non-zero value)
```

nxqbao commented on contracts/ronin/slash-indicator/CreditScore.sol:
```diff
@@ -184,8 +200,9 @@ abstract contract CreditScore is ICreditScore, HasValidatorContract, HasMaintena
-    require(_cutOffPercentage <= _MAX_PERCENTAGE, "CreditScore: invalid cut off percentage config");
+    if (_gainScore > _maxScore) revert ErrInvalidCreditScoreConfig();
+
+    if (_cutOffPercentage > _MAX_PERCENTAGE) revert ErrInvalidCutOffPercentageConfig();
```
```suggestion
    if (_gainScore > _maxScore) revert ErrInvalidCreditScoreConfig();
    if (_cutOffPercentage > _MAX_PERCENTAGE) revert ErrInvalidCutOffPercentageConfig();
```

nxqbao commented on contracts/ronin/staking/CandidateStaking.sol:
```diff
@@ -235,9 +241,8 @@ abstract contract CandidateStaking is BaseStaking, ICandidateStaking, GlobalConf
     if (_amount > _pool.stakingAmount) revert ErrInsufficientStakingAmount();
-    if (_pool.lastDelegatingTimestamp[_requester] + _cooldownSecsToUndelegate > block.timestamp) {
+    if (_pool.lastDelegatingTimestamp[_requester] + _cooldownSecsToUndelegate > block.timestamp)
       revert ErrUnstakeTooEarly();
```
```suggestion
    if (_pool.lastDelegatingTimestamp[_requester] + _cooldownSecsToUndelegate > block.timestamp) {
      revert ErrUnstakeTooEarly();
    }
```

Commit:
format: rename deprecated variable

nxqbao commented on contracts/ronin/Maintenance.sol:
```diff
@@ -78,33 +79,31 @@ contract Maintenance is IMaintenance, HasValidatorContract, Initializable {
+
+    if (!_validator.epochEndingAt(_startedAtBlock - 1)) revert ErrStartBlockOutOfRange();
+
+    if (!_validator.epochEndingAt(_endedAtBlock)) revert ErrEndBlockOutOfRange();
```
Remove blank space among these `if`.
Add `{}` for in-new-line `revert`.

nxqbao commented on contracts/ronin/Maintenance.sol:
```diff
@@ -118,12 +117,12 @@ contract Maintenance is IMaintenance, HasValidatorContract, Initializable {
+
+    if (!checkScheduled(_consensusAddr)) revert ErrUnexistedSchedule();
+
+    if (checkMaintained(_consensusAddr, block.number)) revert ErrAlreadyOnMaintenance();
```
Remove blank space among these if. Add {} for in-new-line revert.

nxqbao commented on contracts/ronin/Maintenance.sol:
```diff
@@ -229,14 +236,10 @@ contract Maintenance is IMaintenance, HasValidatorContract, Initializable {
-    );
+    if (_minMaintenanceDurationInBlock >= _maxMaintenanceDurationInBlock) revert ErrInvalidMaintenanceDurationConfig();
+
+    if (_minOffsetToStartSchedule >= _maxOffsetToStartSchedule) revert ErrInvalidOffsetToStartScheduleConfigs();
```
```suggestion
    if (_minMaintenanceDurationInBlock >= _maxMaintenanceDurationInBlock) revert ErrInvalidMaintenanceDurationConfig();
    if (_minOffsetToStartSchedule >= _maxOffsetToStartSchedule) revert ErrInvalidOffsetToStartScheduleConfigs();
```

nxqbao commented:
PR looks good, saved a huge size.

ducthotran2010 committed:
refactor: add cmt, rename var, update asm

nxqbao committed:
add comment of original code for asm

Commit:
Update contracts/libraries/DeprecatedSlots.sol

Co-authored-by: Bao <quocbao300@gmail.com>

Commit:
Update contracts/libraries/Role.sol

Co-authored-by: Bao <quocbao300@gmail.com>

Commit:
Update contracts/libraries/Role.sol

Co-authored-by: Bao <quocbao300@gmail.com>

Commit:
Update contracts/libraries/Token.sol

Co-authored-by: Bao <quocbao300@gmail.com>

Commit:
Update contracts/ronin/gateway/RoninGatewayV2.sol

Co-authored-by: Bao <quocbao300@gmail.com>

Commit:
Update contracts/ronin/gateway/RoninGatewayV2.sol

Co-authored-by: Bao <quocbao300@gmail.com>

Commit:
Update contracts/ronin/slash-indicator/CreditScore.sol

Co-authored-by: Bao <quocbao300@gmail.com>

Commit:
Update contracts/ronin/staking/CandidateStaking.sol

Co-authored-by: Bao <quocbao300@gmail.com>

Commit:
Update contracts/ronin/Maintenance.sol

Co-authored-by: Bao <quocbao300@gmail.com>

ducthotran2010 commented on contracts/extensions/GovernanceAdmin.sol:
```diff
@@ -24,33 +25,48 @@ abstract contract GovernanceAdmin is CoreGovernance, HasRoninTrustedOrganization
-    _setRoninTrustedOrganizationContract(_roninTrustedOrganizationContract);
-    _setBridgeContract(_bridgeContract);
-  }
+    /** @dev
+     * DOMAIN_SEPARATOR = keccak256(
+     *  abi.encode(
+     *    keccak256("EIP712Domain(string name,string version,bytes32 salt)"),
+     *    keccak256("GovernanceAdmin"), // name hash
+     *    keccak256("2"), // version hash
+     *    keccak256(abi.encode("RONIN_GOVERNANCE_ADMIN", _roninChainId)) // salt
+     *  )
+     *);
+     */
```
Unexpected NatSpec tag "dev" with value in inline assembly.solidity

ducthotran2010 committed:
fix rebase issues

ducthotran2010 commented on contracts/libraries/EmergencyExitBallot.sol:
```diff
@@ -16,16 +16,27 @@ library EmergencyExitBallot {
-        )
-      );
+  ) internal pure returns (bytes32 digest) {
+    /** @dev
+     * return
+     *   keccak256(
+     *     abi.encode(
+     *       EMERGENCY_EXIT_BALLOT_TYPEHASH,
+     *       _consensusAddress,
+     *       _recipientAfterUnlockedFund,
+     *       _requestedAt,
+     *       _expiredAt
+     *     )
+     *   );
+     **/
```
Unexpected NatSpec tag "dev" with value in inline assembly.solidity(6269)

nxqbao committed:
remove dev tag

ducthotran2010 commented on contracts/libraries/Role.sol:
```diff
@@ -0,0 +1,31 @@
+ * @notice Sort role by contract or EOA in parity order.
+ * @dev The 1st (the right bit) most is reserved. EOA addresses are even (`*0`), contract addresses are odd (`*1`)
+ */
+enum Role {
+   /*  0 */ UNKNOWN_0,
+   /*  1 */ UNKNOWN_1,
+   /*  2 */ ADMIN,
+   /*  3 */ PAUSE_ENFORCER_CONTRACT,
+   /*  4 */ COINBASE,
+   /*  5 */ BRIDGE_CONTRACT,
+   /*  6 */ GOVERNOR,
+   /*  7 */ BRIDGE_TRACKING_CONTRACT,
+   /*  8 */ CANDIDATE_ADMIN,
+   /*  9 */ GOVERNANCE_ADMIN_CONTRACT,
+   /* 10 */ WITHDRAWAL_MIGRATOR,
+   /* 11 */ MAINTENANCE_CONTRACT,
+   /* 12 */ BRIDGE_OPERATOR,
+   /* 13 */ SLASH_INDICATOR_CONTRACT,
+   /* 14 */ BLOCK_PRODUCER,
+   /* 15 */ STAKING_VESTING_CONTRACT,
+   /* 16 */ VALIDATOR_CANDIDATE,
+   /* 17 */ VALIDATOR_CONTRACT,
+   /* 18 */ RESERVED_0,
+   /* 19 */ STAKING_CONTRACT,
+   /* 20 */ RESERVED_1,
+   /* 21 */ RONIN_TRUSTED_ORGANIZATION_CONTRACT
+}
```
I don't think we should merge these 2 sets together including EOA roles, and contract types.

nxqbao committed:
Merge branch 'dev' into reduce-contract-size

nxqbao committed:
add storage layout

nxqbao committed:
fix comment

nxqbao committed:
Merge branch 'dev' into reduce-contract-size

Commit:
format: move Role, DeprecatedSlots, CommonErrors to utils

Commit:
feat: change error ErrUnsupportedTarget

Commit:
feat: change error ErrInsufficientGas

Commit:
feat: Remove blank space among these if.
Add {} for in-new-line revert.

Commit:
feat: deleted unused file

Commit:
feat: update pre-commit  & storage layout log

Commit:
format: rename deprecated var

ducthotran2010 committed:
Split Role enums & adaptation

ducthotran2010 commented:
This PR includes heavy changes, and needs some extra attention from the security aspect.

nxqbao commented on test/helpers/utils.ts:
```diff
@@ -56,3 +56,21 @@ export const accessControlRevertStr = (addr: Address, role: string): string =>
 export const compareBigNumbers = (firstBigNumbers: BigNumber[], secondBigNumbers: BigNumber[]) =>
   expect(firstBigNumbers.map((_) => _.toHexString())).deep.equal(secondBigNumbers.map((_) => _.toHexString()));
+
+const ROLES = [
```
Rename to `CONTRACT_TYPES`.

nxqbao commented on contracts/extensions/collections/HasContracts.sol:
```diff
@@ -0,0 +1,80 @@
+   */
+  function _requireContract(ContractType contractType) private view {
+    address expectedCaller = getContract(contractType);
+    if (msg.sender != expectedCaller)
+      revert ErrUnexpectedInternalCall(msg.sig, contractType, expectedCaller, msg.sender);
```
```suggestion
    if (msg.sender != expectedCaller) {
      revert ErrUnexpectedInternalCall(msg.sig, contractType, expectedCaller, msg.sender);
    }
```

nxqbao commented on hardhat.config.ts:
```diff
@@ -140,6 +157,13 @@ const config: HardhatUserConfig = {
+    alphaSort: true,
+    disambiguatePaths: false,
+    runOnCompile: true,
+    outputFile: './logs/contract-code-sizes.log',
```
```suggestion
    outputFile: './logs/contract_code_sizes.log',
```

nxqbao committed:
Apply suggestions from code review

nxqbao committed:
rename var

Commit:
Merge branch 'dev' into reduce-contract-size

nxqbao commented on contracts/libraries/Token.sol:
```diff
@@ -25,19 +38,25 @@ library Token {
    */
-  function hash(Info memory _info) internal pure returns (bytes32) {
-    return keccak256(abi.encode(INFO_TYPE_HASH, _info.erc, _info.id, _info.quantity));
+  function hash(Info memory _info) internal pure returns (bytes32 digest) {
+    assembly {
```
```suggestion
  function hash(Info memory _info) internal pure returns (bytes32 digest) {
    // keccak256(abi.encode(INFO_TYPE_HASH, _info.erc, _info.id, _info.quantity))
    assembly {
```

nxqbao commented on contracts/libraries/Token.sol:
```diff
@@ -25,19 +38,25 @@ library Token {
-        (_info.erc == Standard.ERC721 && _info.quantity == 0),
-      "Token: invalid info"
-    );
+    if (
+      !(_info.erc == Standard.ERC20 && _info.quantity > 0 && _info.id == 0) ||
+      (_info.erc == Standard.ERC721 && _info.quantity == 0)
+    ) revert ErrInvalidInfo();
```
Wrong `!`, please double check this

```suggestion
    if (
      !((_info.erc == Standard.ERC20 && _info.quantity > 0 && _info.id == 0) ||
      (_info.erc == Standard.ERC721 && _info.quantity == 0))
    ) revert ErrInvalidInfo();
```

nxqbao commented on contracts/libraries/Token.sol:
```diff
@@ -61,9 +75,7 @@ library Token {
-    }
+    } else revert ErrUnsupportedStandard();
 
     if (!_success) {
```
Please change the below `string(abi.encodePacked)` into custom error.

https://github.com/axieinfinity/ronin-dpos-contracts/pull/243/files#diff-44e5707d9a5ca48b8e94d327904aa677cce540d6c8d7f465a69e185b7d7db3d2R81-R94

nxqbao commented on contracts/libraries/Token.sol:
```diff
@@ -210,7 +206,14 @@ library Token {
    */
-  function hash(Owner memory _owner) internal pure returns (bytes32) {
-    return keccak256(abi.encode(OWNER_TYPE_HASH, _owner.addr, _owner.tokenAddr, _owner.chainId));
+  function hash(Owner memory _owner) internal pure returns (bytes32 digest) {
+    assembly {
```
```suggestion
  function hash(Owner memory _owner) internal pure returns (bytes32 digest) {
    // keccak256(abi.encode(OWNER_TYPE_HASH, _owner.addr, _owner.tokenAddr, _owner.chainId))
    assembly {
```

nxqbao commented on contracts/mainchain/MainchainGatewayV2.sol:
```diff
@@ -276,28 +290,28 @@ contract MainchainGatewayV2 is WithdrawalLimitation, Initializable, AccessContro
-    require(_receipt.mainchain.chainId == block.chainid, "MainchainGatewayV2: invalid chain id");
+    if (_receipt.kind != Transfer.Kind.Withdrawal) revert ErrInvalidReceiptKind();
+
+    if (_receipt.mainchain.chainId != block.chainid)
+      revert ErrInvalidChainId(msg.sig, _receipt.mainchain.chainId, block.chainid);
```
```suggestion
    if (_receipt.mainchain.chainId != block.chainid) {
      revert ErrInvalidChainId(msg.sig, _receipt.mainchain.chainId, block.chainid);
    }
```

nxqbao commented on contracts/mainchain/MainchainGatewayV2.sol:
```diff
@@ -276,28 +290,28 @@ contract MainchainGatewayV2 is WithdrawalLimitation, Initializable, AccessContro
+
+    if (withdrawalHash[_id] != 0) revert ErrQueryForProcessedWithdrawal();
+
+    if (!(_receipt.info.erc == Token.Standard.ERC721 || !_reachedWithdrawalLimit(_tokenAddr, _quantity)))
+      revert ErrReachedDailyWithdrawalLimit();
```
```suggestion
    if (!(_receipt.info.erc == Token.Standard.ERC721 || !_reachedWithdrawalLimit(_tokenAddr, _quantity))) {
      revert ErrReachedDailyWithdrawalLimit();
    }
```

nxqbao commented on contracts/ronin/staking/BaseStaking.sol:
```diff
@@ -33,25 +36,42 @@ abstract contract BaseStaking is
   }
 
   modifier notPoolAdmin(PoolDetail storage _pool, address _delegator) {
-    if (_pool.admin == _delegator) revert ErrPoolAdminForbidden();
+    _notPoolAdmin(_pool, _delegator);
     _;
   }
```
Please help me rename `notPoolAdmin` into `onlyNotPoolAdmin` for adapting code convention.

nxqbao commented on contracts/ronin/staking/BaseStaking.sol:
```diff
@@ -33,25 +36,42 @@ abstract contract BaseStaking is
     _;
   }
 
   modifier poolIsActive(address _poolAddr) {
```
Please help me rename `poolIsActive` into `onlyActivePool` for adapting code convention.

Commit:
feat: remove expectedCaller for ErrUnexpectedInternalCall

Commit:
Update contracts/libraries/Token.sol

Co-authored-by: Bao <quocbao300@gmail.com>

Commit:
Update contracts/libraries/Token.sol

Co-authored-by: Bao <quocbao300@gmail.com>

Commit:
Update contracts/mainchain/MainchainGatewayV2.sol

Co-authored-by: Bao <quocbao300@gmail.com>

Commit:
Update contracts/mainchain/MainchainGatewayV2.sol

Co-authored-by: Bao <quocbao300@gmail.com>

Commit:
fix: fix validate logic bug

Commit:
format: rename notPoolAdmin modifier

Commit:
feat: update contract size log

nxqbao reviewed (approved):
Please update the table of changes in the PR.

Commit:
feat: enable storage slot migration

Commit:
feat: update contract size log

Commit:
feat: add delete operation to deprecated slot value

nxqbao commented on contracts/libraries/Token.sol:
```diff
@@ -123,20 +117,7 @@ library Token {
-        )
-      );
-    }
+    if (!_success) revert ErrTokenCouldNotTransferFrom(_info, address(this), _to, _token);
```
Should declare a new error `ErrTokenCouldNotTransfer` to distinguish from `ErrTokenCouldNotTransferFrom`.

nxqbao committed:
fix: split to ErrTokenCouldNotTransfer error

nxqbao committed:
husky: add contract size log

nxqbao committed:
Merge branch 'dev' into reduce-contract-size

nxqbao committed:
fix yarn.lock

Commit:
Merge branch 'dev' into reduce-contract-size
